### ISSUE TITLE
Add completed callback map on `SpriteAnimationGroupComponent`

### ISSUE THREAD

**Issue by oloshe** [flame-engine/flame#1654]
Sometimes we use `SpriteAnimationGroupComponent` to control a sprite character to act something, it's convenient and useful. However, when I want to add some logic after some animation completed, it need to do some inconvenient steps, it was like:

```dart
SpriteAnimationGroupComponent<T> comp;

// ...

comp.animation!.onComplete = onCompleted;

// ...

void onCompleted() {
    // do my logic

    // reset onComplete to null
    comp.animation!.onComplete = null;
}
```

# Problem to solve
It may make it easier and convenient to invoke each status's callback or "hook",  such as reset my status or switch to other animation.

# Proposal

The way I solve it is extends it and add a field just like the `removeOnFinish`:

```dart
class SpriteAnimationGroupCallbackComponent<T> extends SpriteAnimationGroupComponent<T> {
  final Map<T, void Function()> onFinish;
  // ...

  @override
  void update(double dt) {
    super.update(dt);
    if (onFinish[current] != null && (animation?.done() ?? false)) {
      onFinish[current]!.call();
    }
  }
}
```

But it's seem to no need to create an new class to do this trivial thing, maybe add it on `SpriteAnimationGroupComponent` is more suitable.

**Comment by spydon** [comment]
Sounds like a good addition to me, do you want to put up a PR for it?

**Comment by oloshe** [comment]
alright, i'll create a pr later

**Comment by munsterlander** [comment]
I am confused on why this is needed.  As a group component, it is just implementing `SpriteAnimation` underneath the `SpriteAnimationComponent` that it wraps.  Those individual animations already have `onStart`, `onFrame`, and `onComplete` callbacks.  As a group component, there is not an end of the group, but an end of the individual animation that is current.  So in your group component - at least this is how I am doing it - you just need:

```
    final testAnimation = testSpriteSheet.createAnimation(row: 0, stepTime: 1, loop: false, to: 8);
    testAnimation.onComplete = (){
    //whatever
    };
    animations = {
      TestAnimationGroupState.testState: testAnimation,
      ....
};
etc...
```

In short, this was the whole reason we added the other callbacks, so why is a wrapper component callback necessary as the group component doesn't really have an end as you can change current to a different animation? 

Again, I could be completely lost which is a common occurrence.

**Comment by oloshe** [comment]
Thanks for @munsterlander 
You answered my doubts. I wrapped so many that I forgot they could be defined individually.

**Comment by oloshe** [comment]
Maybe this issue can be closed. This solution is more gentlemanly


> ```
>     final testAnimation = testSpriteSheet.createAnimation(row: 0, stepTime: 1, loop: false, to: 8);
>     testAnimation.onComplete = (){
>     //whatever
>     };
>     animations = {
>       TestAnimationGroupState.testState: testAnimation,
>       ....
> };
> etc...
> ```

**Comment by munsterlander** [comment]
Glad I could help! Again, I'm still learning flame so maybe there is a need, but you can definitely accomplish your task with everything as is I think.

**Comment by oloshe** [comment]
As you said, people may get lost in its two completed callbacks. However, whether it is a need, it may needs to be reviewed again.

**Comment by munsterlander** [comment]
I am somewhat confused by the on finish for the group component to begin with. What finish? If you have 3 animations, is that supposed to be called at the end of every animation as a global non-specific callback?  I don't know, like you said, the reason for 2 callbacks.

**Comment by spydon** [comment]
We should update the docs for `SpriteAnimationGroupComponent` with an example of how to use those callbacks, or link to the `SpriteAnimation` docs from there at least.

**Comment by munsterlander** [comment]
Ok, I can probably take that on this weekend.

**Comment by oloshe** [comment]
great

**Comment by munsterlander** [comment]
Done, but something wonky is afoot in that PR.

**Comment by munsterlander** [comment]
We can close this one now.
